### PR TITLE
[Appsec] Fix posting json missing with HttpStreamRequest: missing header / reset streams

### PR DESCRIPTION
**Pull request by anna-git**
Changes proposed in this pull request:
- Adding content type application/json header for IApiWebRequest pipes implementation (HttpStreamRequest)
- Seek stream to 0 to avoid truncated json
- Dispose response in Sender.cs 
- Log headers, payload (available only for HttpStreamRequest, ApiWebRequest cant read the request stream twice, HttpClientRequest is disposing it when posting)

> Fixes https://datadoghq.atlassian.net/browse/APPSEC-1447

@DataDog/apm-dotnet

### CODE REVIEW

**Review by robertpi — APPROVED** [review]

File: tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs [hunk 1]
```diff
@@ -66,19 +66,32 @@ public async Task<IApiResponse> PostAsJsonAsync(IEvent events, JsonSerializer se
...
+                when (exception.Status == WebExceptionStatus.ProtocolError && exception.Response != null)
+            {
+                // If the exception is caused by an error status code, ignore it and let the caller handle the result
+                return new ApiWebResponse((HttpWebResponse)exception.Response);
+            }
+        }
+
+        public Task<string> RequestContent()
+        {
+            var sb = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (anna-git) [reply tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs]: Fixed in [6f709ea515c4471d62b841d143d588df306ee0eb](https://github.com/DataDog/dd-trace-dotnet/pull/1778/commits/6f709ea515c4471d62b841d143d588df306ee0eb)

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1778 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) will **not change** line coverage
:heavy_check_mark: Merging #1778 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) will **not change** branch coverage
:heavy_check_mark: Merging #1778 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) will will **decrease** complexity by `8`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) | #1778       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9262` / `14709` | `9246` / `14721` |          |
| Lines %   | `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4122` / `6418` | `4118` / `6422` |          |
| Branches %| `0.64%`      | `0.64%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7570`      | `7562`        |  `-8`  :heavy_check_mark:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75153/artifacts?artifactName=Code%20Coverage%20Report_75153&api-version=6.0&%24format=zip)
* [For this PR #1778](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75171/artifacts?artifactName=Code%20Coverage%20Report_75171&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/247da284d4b40f5dfa65c36e43b347a527906454) | #1778       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.64%`      | `0.64%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7570`      | `7562`       |  `-8` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Agent.Transports.HttpStreamRequest](https://github.com/DataDog/dd-trace-dotnet/tree/7d7284d513d80b68f014aa60f4bd25060bf7fe25/tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs) | `-0.25%` :no_entry: | `-0.33%` :no_entry:   | `4` :no_entry: |
| [Datadog.Trace.Agent.Transports.ApiWebRequest](https://github.com/DataDog/dd-trace-dotnet/tree/7d7284d513d80b68f014aa60f4bd25060bf7fe25/tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs) | `-0.11%` :no_entry: | `0.17%` :heavy_check_mark:   | `-1` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.Helpers.AsyncHelper](https://github.com/DataDog/dd-trace-dotnet/tree/7d7284d513d80b68f014aa60f4bd25060bf7fe25/tracer/src/Datadog.Trace/ClrProfiler/Helpers/AsyncHelper.cs) | `0.00%` :heavy_check_mark: | `-0.08%` :no_entry:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75153/artifacts?artifactName=Code%20Coverage%20Report_75153&api-version=6.0&%24format=zip)
* [For this PR #1778](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75171/artifacts?artifactName=Code%20Coverage%20Report_75171&api-version=6.0&%24format=zip)

File: tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs [hunk 2]
```diff
@@ -41,11 +41,14 @@ internal async Task Send(IEnumerable<IEvent> events)
...
-            if (response.StatusCode != 200 && response.StatusCode != 202)
+            using (var response = await request.PostAsJsonAsync(batch, _serializer))
             {
-                var responseText = await response.ReadAsStringAsync();
-                Log.Warning($"AppSec event not correctly sent to backend {response.StatusCode} with response {responseText}");
+                if (response.StatusCode != 200 && response.StatusCode != 202)
+                {
+                    var statusCode = response.StatusCode;
+                    var responseText = await response.ReadAsStringAsync();
+                    Log.Warning($"AppSec event not correctly sent to backend {statusCode} with response {responseText}, request content was: {await request.RequestContent()}");
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs:50]: nit: should use structured logging
```suggestion
                    Log.Warning("AppSec event not correctly sent to backend {StatusCode} with response {ResponseText}, request content was: {Content}", statusCode, responseText, await request.RequestContent());
```

**Review by andrewlock — COMMENTED** [review]
I'm not a fan of adding this to the `IApiRequest` API, though I can see why you did it. Given it's only valid to be called after you call `Post*`, it's only used by app sec (and only for logging), and only even valid at all in one of three implementations, I wonder if this should be an "internal" implementation in `HttpStreamRequest.PostAsJsonAsync()`? _Especially_ as you already have a memory stream with the contents that you're sending at that point

File: tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs (hunk 2 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs]: Thanks for the feedback! So I removed the implementation, and now all 3 impl. can log the payload, I just log directly in the method if the response status code is in error.

File: tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs [hunk 3]
```diff
@@ -41,12 +41,7 @@ internal async Task Send(IEnumerable<IEvent> events)
             };
             var request = _apiRequestFactory.Create(_uri);
             request.AddHeader(AppSecHeaderKey, AppSecHeaderValue);
-            var response = await request.PostAsJsonAsync(batch, _serializer);
-            if (response.StatusCode != 200 && response.StatusCode != 202)
-            {
-                var responseText = await response.ReadAsStringAsync();
-                Log.Warning($"AppSec event not correctly sent to backend {response.StatusCode} with response {responseText}");
-            }
+            return request.PostAsJsonAsync(batch, _serializer);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs:44]: Not passionate about it, but you _might_ still want to await this one to get a better stack trace if it errors?

File: tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs [hunk 4]
```diff
@@ -48,8 +52,22 @@ public async Task<IApiResponse> PostAsJsonAsync(IEvent events, JsonSerializer se
             {
                 serializer.Serialize(writer, events);
                 await writer.FlushAsync();
+                await memoryStream.FlushAsync();
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs:55]: I think these should have `ConfigureAwait(false)` too?

File: tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs [hunk 5]
```diff
@@ -63,10 +81,9 @@ private async Task<IApiResponse> PostSegmentAsync(ArraySegment<byte> segment)
             using (var bidirectionalStream = _streamFactory.GetBidirectionalStream())
             {
                 var content = new BufferContent(segment);
-                var request = new HttpRequest("POST", _uri.Host, _uri.PathAndQuery, _headers, content);
-
+                _request = new HttpRequest("POST", _uri.Host, _uri.PathAndQuery, _headers, content);
                 // send request, get response
-                var response = await _client.SendAsync(request, bidirectionalStream, bidirectionalStream).ConfigureAwait(false);
+                var response = await _client.SendAsync(_request, bidirectionalStream, bidirectionalStream).ConfigureAwait(false);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs:86]: Thanks for the update @anna-git 🙂 This mutable `_request` variable still bothers me a bit though, as it's only non-null/valid to access at certain points. Could we change the signature of the private `PostSegmentAsync` method to return _both_ the `IApiResponse` and the `HttpRequest`? That would remove  the need for the field

File: tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs [hunk 6]
```diff
@@ -62,21 +63,44 @@ public async Task<IApiResponse> PostAsJsonAsync(IEvent events, JsonSerializer se
 
             using (var requestStream = await _request.GetRequestStreamAsync().ConfigureAwait(false))
             {
-                using (var writer = new JsonTextWriter(new StreamWriter(requestStream)))
+                Task WriteStream(Stream stream)
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs:66]: Should we make this a static local function to avoid the closure?

**Review by andrewlock — COMMENTED** [review]
Thanks for the update 🙂 Just a couple more questions!

File: tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs (hunk 6 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs]: Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1778/commits/62c946f8890c5547eb463e90479a50f783e369bf

File: tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs (hunk 5 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs]: Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1778/commits/62c946f8890c5547eb463e90479a50f783e369bf

File: tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs (hunk 4 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs]: Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1778/commits/62c946f8890c5547eb463e90479a50f783e369bf

File: tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs (hunk 3 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs]: I thought if you await at caller level, the exception bubbles up fine.. Although PostAsJsonAsync returns a Task<IApiResponse> and the caller doesn't use the result yet, just awaiting it. So I'm not sure in this case... 😅 
Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1778/commits/62c946f8890c5547eb463e90479a50f783e369bf

Comment (andrewlock) [reply tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs]: Yeah the exception bubbles up fine either way, but you get a slightly screwy stack trace. For example if `PostAsJsonAsync()` throws an exception, and it isn't awaited here, `Send()` won't appear on the stack trace. But it's less efficient to add the await, so it's a trade off 😉

File: tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs [hunk 7]
```diff
@@ -62,21 +63,44 @@ public async Task<IApiResponse> PostAsJsonAsync(IEvent events, JsonSerializer se
...
-                        var httpWebResponse = (HttpWebResponse)await _request.GetResponseAsync().ConfigureAwait(false);
-                        return new ApiWebResponse(httpWebResponse);
+                        serializer.Serialize(writer, events);
+                        return writer.FlushAsync();
                     }
-                    catch (WebException exception)
-                        when (exception.Status == WebExceptionStatus.ProtocolError && exception.Response != null)
+                }
+
+                await WriteStream(requestStream, serializer, events);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs:76]: I spotted some more `ConfigureAwait(false)` 😅 
```suggestion
                await WriteStream(requestStream, serializer, events).ConfigureAwait(false);
```

File: tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs [hunk 8]
```diff
@@ -62,21 +63,44 @@ public async Task<IApiResponse> PostAsJsonAsync(IEvent events, JsonSerializer se
...
-                        return new ApiWebResponse((HttpWebResponse)exception.Response);
+                        var sb = Util.StringBuilderCache.Acquire(0);
+                        foreach (var item in _request.Headers)
+                        {
+                            sb.Append($"{item}: {_request.Headers[item.ToString()]} ");
+                            sb.Append(", ");
+                        }
+
+                        using var ms = new MemoryStream();
+                        await WriteStream(ms, serializer, events);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs:91]:
```suggestion
                        await WriteStream(ms, serializer, events).ConfigureAwait(false);
```

File: tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs [hunk 9]
```diff
@@ -41,12 +40,7 @@ internal async Task Send(IEnumerable<IEvent> events)
             };
             var request = _apiRequestFactory.Create(_uri);
             request.AddHeader(AppSecHeaderKey, AppSecHeaderValue);
-            var response = await request.PostAsJsonAsync(batch, _serializer);
-            if (response.StatusCode != 200 && response.StatusCode != 202)
-            {
-                var responseText = await response.ReadAsStringAsync();
-                Log.Warning($"AppSec event not correctly sent to backend {response.StatusCode} with response {responseText}");
-            }
+            await request.PostAsJsonAsync(batch, _serializer);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs:43]:
```suggestion
            await request.PostAsJsonAsync(batch, _serializer).ConfigureAwait(false);
```

**Review by andrewlock — APPROVED** [review]

File: tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs (hunk 3 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/AppSec/Transports/Sender.cs]: oh ok thanks for the explanation :) didn't know the intermediate asynchronous methods that don't await wouldn't appear!

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1778 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/1c2bd53c6463c809b973055d7ad67244395a779e) will **not change** line coverage
:heavy_check_mark: Merging #1778 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/1c2bd53c6463c809b973055d7ad67244395a779e) will **not change** branch coverage
:no_entry: Merging #1778 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/1c2bd53c6463c809b973055d7ad67244395a779e) will will **increase** complexity by `15`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/1c2bd53c6463c809b973055d7ad67244395a779e) | #1778       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9436` / `14997` | `9431` / `15031` |          |
| Lines %   | `63%`      | `63%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4163` / `6496` | `4163` / `6510` |          |
| Branches %| `64%`      | `64%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7657`      | `7672`        |  `15`  :no_entry:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75796&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75796/artifacts?artifactName=Code%20Coverage%20Report_75796&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1778](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75807&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75807/artifacts?artifactName=Code%20Coverage%20Report_75807&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/1c2bd53c6463c809b973055d7ad67244395a779e) | #1778       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `63%`      | `63%`       |  `0%` :heavy_check_mark:  |
| Branches %| `64%`      | `64%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7657`      | `7672`       |  `15` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Agent.Transports.HttpStreamRequest](https://github.com/DataDog/dd-trace-dotnet/tree/eca53754c90302ec7af022b838b9ccbfc1c91fb6/tracer/src/Datadog.Trace/Agent/Transports/HttpStreamRequest.cs) | `-22%` :no_entry: | `-33%` :no_entry:   | `4` :no_entry: |
| [Datadog.Trace.Agent.Transports.ApiWebRequest](https://github.com/DataDog/dd-trace-dotnet/tree/eca53754c90302ec7af022b838b9ccbfc1c91fb6/tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs) | `-21%` :no_entry: | `-21%` :no_entry:   | `11` :no_entry: |
| [Datadog.Trace.Agent.Transports.HttpClientRequest](https://github.com/DataDog/dd-trace-dotnet/tree/eca53754c90302ec7af022b838b9ccbfc1c91fb6/tracer/src/Datadog.Trace/Agent/Transports/HttpClientRequest.cs) | `-12%` :no_entry: | `-100%` :no_entry:   | `4` :no_entry: |
| [Datadog.Trace.RuntimeMetrics.RuntimeMetricsWriter](https://github.com/DataDog/dd-trace-dotnet/tree/eca53754c90302ec7af022b838b9ccbfc1c91fb6/tracer/src/Datadog.Trace/RuntimeMetrics/RuntimeMetricsWriter.cs) | `0%` :heavy_check_mark: | `8%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Logging.LogRateLimiter](https://github.com/DataDog/dd-trace-dotnet/tree/eca53754c90302ec7af022b838b9ccbfc1c91fb6/tracer/src/Datadog.Trace/Logging/LogRateLimiter.cs) | `0%` :heavy_check_mark: | `17%` :heavy_check_mark:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75796&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75796/artifacts?artifactName=Code%20Coverage%20Report_75796&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1778](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75807&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75807/artifacts?artifactName=Code%20Coverage%20Report_75807&api-version=6.0&%24format=zip)

**Pull request merged**